crates/client/rpc/src/lib.rs

```diff
@@ -681,13 +708,128 @@ where
     }
 
     /// Returns all events matching the given filter
-    async fn get_events(
-        &self,
-        filter: EventFilter,
-        continuation_token: Option<String>,
-        chunk_size: u64,
-    ) -> RpcResult<EventsPage> {
-        todo!("Not implemented")
+    async fn get_events(&self, filter: EventFilterWithPage) -> RpcResult<EventsPage> {
+        let mut continuation_token = match filter.result_page_request.continuation_token {
+            Some(token) => token.parse::<usize>().map_err(|e| {
+                error!("Failed to parse continuation token: {:?}", e);
+                StarknetRpcApiError::InvalidContinuationToken
+            })?,
+            None => 0usize,
+        };
+        let filter_address = filter.event_filter.address.map(Felt252Wrapper::from);
+        let keys = filter.event_filter.keys.unwrap_or_default();
+        let chunk_size = filter.result_page_request.chunk_size;
+
+        if keys.len() > MAX_EVENTS_KEYS {
+            return Err(StarknetRpcApiError::TooManyKeysInFilter.into());
+        }
+        if chunk_size > MAX_EVENTS_CHUNK_SIZE {
+            return Err(StarknetRpcApiError::PageSizeTooBig.into());
+        }
+
+        // Get the substrate block numbers for the requested range
+        let latest_block =
+            self.substrate_block_number_from_starknet_block(BlockId::Tag(BlockTag::Latest)).map_err(|e| {
+                error!("'{e}'");
+                StarknetRpcApiError::BlockNotFound
+            })?;
+        let mut current_block = self
+            .substrate_block_number_from_starknet_block(filter.event_filter.from_block.unwrap_or(BlockId::Number(0)))
+            .map_err(|e| {
+                error!("'{e}'");
+                StarknetRpcApiError::BlockNotFound
+            })?;
+        let to_block = self
+            .substrate_block_number_from_starknet_block(
+                filter.event_filter.to_block.unwrap_or(BlockId::Tag(BlockTag::Latest)),
+            )
+            .map_err(|e| {
+                error!("'{e}'");
+                StarknetRpcApiError::BlockNotFound
+            })?;
+
+        // Verify that the requested range is valid
+        if current_block > to_block {
+            return Ok(EventsPage { events: vec![], continuation_token: None });
+        }
+
+        let to_block = if latest_block > to_block { to_block } else { latest_block };
+
+        let mut filtered_events = vec![];
+        let mut index = 0;
+        while current_block <= to_block {
+            let substrate_block_hash =
+                self.substrate_block_hash_from_starknet_block(BlockId::Number(current_block)).map_err(|e| {
```

Comment:
can we call directly on `self.client.hash.` as we already know we're dealing with a `BlockId::Number` and save an inner match?